File: agents/src/middleware/jwt.test.ts
/**
 * Copyright (c) 2026, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { test } from "node:test";
import assert from "node:assert/strict";
import { createServer, type Server } from "node:http";
import { AddressInfo } from "node:net";
import { generateKeyPair, exportJWK, SignJWT, type JWK, type KeyLike } from "jose";
import express from "express";
import { jwtAuthMiddleware } from "./jwt.js";

interface TestContext {
  jwksServer: Server;
  appServer: Server;
  appPort: number;
  privateKey: KeyLike | Uint8Array;
  jwk: JWK;
  jwksUrl: string;
}

async function listen(server: Server): Promise<number> {
  return new Promise((resolve) => {
    server.listen(0, "127.0.0.1", () => {
      resolve((server.address() as AddressInfo).port);
    });
  });
}

async function close(server: Server): Promise<void> {
  return new Promise((resolve) => server.close(() => resolve()));
}

async function setup(): Promise<TestContext> {
  const { publicKey, privateKey } = await generateKeyPair("RS256");
  const jwk = await exportJWK(publicKey);
  jwk.kid = "test-key-1";
  jwk.alg = "RS256";
  jwk.use = "sig";

  const jwksServer = createServer((_req, res) => {
    res.setHeader("Content-Type", "application/json");
    res.end(JSON.stringify({ keys: [jwk] }));
  });
  const jwksPort = await listen(jwksServer);
  const jwksUrl = `http://127.0.0.1:${jwksPort}/jwks`;

  const app = express();
  app.use(
    jwtAuthMiddleware({
      jwksUrl,
      issuer: "test-iss",
      audience: "test-aud",
    }),
  );
  app.get("/secure", (_req, res) => {
    res.status(200).json({ ok: true });
  });
  const appServer = createServer(app);
  const appPort = await listen(appServer);

  return { jwksServer, appServer, appPort, privateKey, jwk, jwksUrl };
}

async function teardown(ctx: TestContext): Promise<void> {
  await close(ctx.appServer);
  await close(ctx.jwksServer);
}

test("jwtAuthMiddleware: accepts valid token", async () => {
  const ctx = await setup();
  try {
    const token = await new SignJWT({})
      .setProtectedHeader({ alg: "RS256", kid: "test-key-1" })
      .setIssuer("test-iss")
      .setAudience("test-aud")
      .setIssuedAt()
      .setExpirationTime("1h")
      .sign(ctx.privateKey);

    const resp = await fetch(`http://127.0.0.1:${ctx.appPort}/secure`, {
      headers: { Authorization: `Bearer ${token}` },
    });
    assert.equal(resp.status, 200);
    const body = await resp.json();
    assert.deepEqual(body, { ok: true });
  } finally {
    await teardown(ctx);
  }
});

test("jwtAuthMiddleware: rejects missing Authorization header", async () => {
  const ctx = await setup();
  try {
    const resp = await fetch(`http://127.0.0.1:${ctx.appPort}/secure`);
    assert.equal(resp.status, 401);
    const wwwAuth = resp.headers.get("www-authenticate");
    assert.ok(wwwAuth?.startsWith('Bearer realm="asdlc"'), `WWW-Authenticate=${wwwAuth}`);
  } finally {
    await teardown(ctx);
  }
});

test("jwtAuthMiddleware: rejects malformed Authorization header", async () => {
  const ctx = await setup();
  try {
    const resp = await fetch(`http://127.0.0.1:${ctx.appPort}/secure`, {
      headers: { Authorization: "NotBearer xyz" },
    });
    assert.equal(resp.status, 401);
    const wwwAuth = resp.headers.get("www-authenticate");
    assert.ok(wwwAuth?.includes('error="invalid_token"'), `WWW-Authenticate=${wwwAuth}`);
  } finally {
    await teardown(ctx);
  }
});

test("jwtAuthMiddleware: rejects wrong audience", async () => {
  const ctx = await setup();
  try {
    const token = await new SignJWT({})
      .setProtectedHeader({ alg: "RS256", kid: "test-key-1" })
      .setIssuer("test-iss")
      .setAudience("wrong-aud")
      .setExpirationTime("1h")
      .sign(ctx.privateKey);

    const resp = await fetch(`http://127.0.0.1:${ctx.appPort}/secure`, {
      headers: { Authorization: `Bearer ${token}` },
    });
    assert.equal(resp.status, 401);
  } finally {
    await teardown(ctx);
  }
});

test("jwtAuthMiddleware: rejects wrong issuer", async () => {
  const ctx = await setup();
  try {
    const token = await new SignJWT({})
      .setProtectedHeader({ alg: "RS256", kid: "test-key-1" })
      .setIssuer("wrong-iss")
      .setAudience("test-aud")
      .setExpirationTime("1h")
      .sign(ctx.privateKey);

    const resp = await fetch(`http://127.0.0.1:${ctx.appPort}/secure`, {
      headers: { Authorization: `Bearer ${token}` },
    });
    assert.equal(resp.status, 401);
  } finally {
    await teardown(ctx);
  }
});

test("jwtAuthMiddleware: rejects expired token", async () => {
  const ctx = await setup();
  try {
    // exp in the past
    const token = await new SignJWT({})
      .setProtectedHeader({ alg: "RS256", kid: "test-key-1" })
      .setIssuer("test-iss")
      .setAudience("test-aud")
      .setIssuedAt(Math.floor(Date.now() / 1000) - 7200)
      .setExpirationTime(Math.floor(Date.now() / 1000) - 60)
      .sign(ctx.privateKey);

    const resp = await fetch(`http://127.0.0.1:${ctx.appPort}/secure`, {
      headers: { Authorization: `Bearer ${token}` },
    });
    assert.equal(resp.status, 401);
  } finally {
    await teardown(ctx);
  }
});

test("jwtAuthMiddleware: rejects unknown kid", async () => {
  const ctx = await setup();
  try {
    // Sign with the right key but lie about kid → JWKS lookup fails.
    const token = await new SignJWT({})
      .setProtectedHeader({ alg: "RS256", kid: "no-such-kid" })
      .setIssuer("test-iss")
      .setAudience("test-aud")
      .setExpirationTime("1h")
      .sign(ctx.privateKey);

    const resp = await fetch(`http://127.0.0.1:${ctx.appPort}/secure`, {
      headers: { Authorization: `Bearer ${token}` },
    });
    assert.equal(resp.status, 401);
  } finally {
    await teardown(ctx);
  }
});

File: services/aep-api/internal/platform/auth/jwt_test.go
// Copyright (c) 2026, WSO2 LLC. (https://www.wso2.com).
//
// WSO2 LLC. licenses this file to you under the Apache License,
// Version 2.0 (the "License"); you may not use this file except
// in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package auth

import (
	"reflect"
	"testing"

	"github.com/wso2/aep/aep-api/internal/authz"
)

func TestClaimsPermissions(t *testing.T) {
	cases := []struct {
		name  string
		scope string
		want  []authz.Permission
	}{
		{
			name:  "mixed known and unknown scopes",
			scope: "openid profile ae:build ae:build-view system",
			want:  []authz.Permission{authz.PermissionBuild, authz.PermissionBuildView},
		},
		{
			name:  "no ae: scopes at all",
			scope: "openid profile system",
			want:  nil,
		},
		{
			name:  "empty scope",
			scope: "",
			want:  nil,
		},
		{
			name:  "whitespace-only scope",
			scope: "   \t  ",
			want:  nil,
		},
		{
			name:  "extra whitespace between scopes",
			scope: "  ae:build   ae:skill-config  ",
			want:  []authz.Permission{authz.PermissionBuild, authz.PermissionSkillConfig},
		},
		{
			name:  "a single unknown ae:-shaped token is not a permission",
			scope: "ae:not-a-real-permission",
			want:  nil,
		},
		{
			name:  "duplicate scope entries are not deduplicated",
			scope: "ae:build ae:build",
			want:  []authz.Permission{authz.PermissionBuild, authz.PermissionBuild},
		},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			claims := &Claims{Scope: tc.scope}
			got := claims.Permissions()
			if !reflect.DeepEqual(got, tc.want) {
				t.Fatalf("Permissions() = %v, want %v", got, tc.want)
			}
		})
	}
}

func TestClaimsPermissions_NilReceiver(t *testing.T) {
	var claims *Claims
	if got := claims.Permissions(); got != nil {
		t.Fatalf("Permissions() on a nil *Claims = %v, want nil", got)
	}
}
